internal/cmd/gocdk/launch.go

```diff
@@ -364,6 +377,75 @@ func (crl *cloudRunLauncher) dockerPush(ctx context.Context, imageName string) e
 	return nil
 }
 
+// cloudRunImage computes the image reference needed to launch the given local
+// image reference on gcr.io and the launch specifier. If the returned string
+// is equal to localImage, then no retagging is necessary before pushing.
+func cloudRunImage(localImage string, launchSpecifier map[string]interface{}) (string, error) {
+	name, tag, digest := parseImageRef(localImage)
+	if tag == ":" {
+		return "", xerrors.Errorf("determine image name for Cloud Run: empty tag in %q", localImage)
+	}
+	if specName := specifierStringValue(launchSpecifier, "image_name"); specName != "" {
+		// First, use image name from launch specifier if present.
+		if !isGCRName(specName) {
+			return "", xerrors.Errorf("determine image name for Cloud Run: launch specifier image_name = %q, not a gcr.io name", specName)
+		}
+		name = specName
+	} else if !isGCRName(name) {
+		// Otherwise, if the image name does not have a gcr.io prefix, then prepend it.
+		project := specifierStringValue(launchSpecifier, "project_id")
+		if project == "" {
+			return "", xerrors.New("determine image name for Cloud Run: launch specifier project_id empty")
+		}
+		// TODO(light): This can be wrong for ORG:PROJECT project IDs, but those
+		// are deprecated anyway.
+		name = "gcr.io/" + project + "/" + name
+	}
+	return name + tag + digest, nil
+}
+
+// parseImageRef parses a Docker image reference, as documented in
```

Comment:
Isn't there a library that does this kind of stuff?
E.g.,
  https://github.com/docker/distribution/blob/master/reference/reference.go
  https://github.com/mafintosh/docker-parse-image

Reply:
Yes, although as noted, this one intentionally implements a looser version of that spec. I did verify that this would not add too many "go.mod-finding" deps, so what you're saying is an option. Overall, I don't think there's much benefit in using a separate library for this single function.

(The second link is a JavaScript library.)